src/components/bulk-email-tool/bulk-email-task-manager/BulkEmailContentHistory.jsx

```diff
@@ -1,41 +1,241 @@
+   */
+  const renderError = () => (
+    <div>
+      <Alert variant="danger">
+        <p className="font-weight-bold">
+          {intl.formatMessage(messages.errorFetchingData)}
+        </p>
+      </Alert>
+    </div>
+  );
+
+  /**
+   * Render function for the email content history table. If there is no data to display in our table we will render
+   * this informative message instead.
+   */
+  const renderEmpty = () => (
+    <div className="pt-1">
+      <Alert variant="warning">
+        <p className="font-weight-bold">
+          {intl.formatMessage(messages.noEmailData)}
+        </p>
+      </Alert>
+    </div>
+  );
+
+  /**
+   * Renders a modal that will display the contents of a single historical email message sent via the bulk course email
+   * tool to a user.
+   */
+  const renderMessageModal = () => (
+    <div>
+      <Modal
+        open={isMessageModalOpen}
+        title=""
+        body={(
+          <div>
+            <div className="d-flex flex-row">
+              <p>
+                {intl.formatMessage(messages.modalMessageSubject)}
+              </p>
+              <p className="pl-2">
+                {messageContent.subject}
+              </p>
+            </div>
+            <div className="d-flex flex-row">
+              <p>
+                {intl.formatMessage(messages.modalMessageSentBy)}
+              </p>
+              <p className="pl-2">
+                {messageContent.requester}
+              </p>
+            </div>
+            <div className="d-flex flex-row">
+              <p>
+                {intl.formatMessage(messages.modalMessageTimeSent)}
+              </p>
+              <p className="pl-2">
+                {messageContent.created}
+              </p>
+            </div>
+            <div className="d-flex flex-row">
+              <p>
+                {intl.formatMessage(messages.modalMessageSentTo)}
+              </p>
+              <p className="pl-2">
+                {messageContent.sent_to}
+              </p>
+            </div>
+            <hr className="py-2" />
+            <div>
+              <p>
+                {intl.formatMessage(messages.modalMessageBody)}
+              </p>
+              <div dangerouslySetInnerHTML={{ __html: messageContent.email.html_message }} />
+            </div>
+          </div>
+        )}
+        onClose={() => setIsMessageModalOpen(false)}
+      />
+    </div>
+  );
+
+  /**
+   * Render function for the email content history table. This function is responsible for displaying data inside of
+   * the table when the `Show Sent Email History` button is pressed on the page.
+   */
+  const renderTable = () => {
+    // Do a little data manipulation to make it easier to display what we want in the table. Pull the email subject out
+    // of the email data. Transforms the `sent_to` array to a string for easier display in our table.
+    const tableData = emailHistoryData.map((item) => ({
+      ...item,
```

Comment:
What other key value pairs come out of this spread? What is the item arg an instance of? Is it documented somewhere? (fyi using jsdoc would potentially answer some of these questions)

Reply:
The `item` is referring to the individual object instance I am interacting with as I am looping through the `emailHistoryData` object. The `emailHistoryData` is an array of  previously sent email messages we got back from our REST API call to edx-platform. I needed to do some data transformation to make things fit the way the Paragon `DataTable` component wants.

We want to display the subject of the message in our data table but this data was buried a level deep in an inner object. When we define the columns of a DataTable, it wants an `accessor` key to get to the data in our objects. This piece of code is pulling that `subject` field out of the inner object and copying it to the top level of the object so it can be displayed in the table.

Before transformation it looks something like...
```
{ 
	created: "Dec 07, 2021 at 18:06 UTC"
	email: {subject: 'asdasdas', html_message: '<p>sfafasfdasd</p>\n<style></style>', id: '50'}
	number_sent: "1 sent"
	requester: "edx"
	sent_to: (2) ['Myself', 'All students']
}
```
and after the transformation it looks like...
```
{ 
	created: "Dec 07, 2021 at 18:06 UTC"
	email: {subject: 'asdasdas', html_message: '<p>sfafasfdasd</p>\n<style></style>', id: '50'}
	number_sent: "1 sent"
	requester: "edx"
	sent_to: "Myself, All students"
	subject: "asdasdas"
}
```

(I also fiddled with the `sent_to` data and transformed it from an array to a String. This was to fix another UI bug in the table. Keeping it an array will display `MyselfAll Students` (no spaces between entries) in the column. 

Hope this makes sense! Let me know if you have further questions.